---
title: "ENCODE API - Metadata extraction"
author: 
- name: "Guillermo Rocamora PÃ©rez"
  affiliation: UCL
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  bookdown::html_document2:
    figure_caption: yes
    code_folding: show
    theme: paper
    highlight: kate
    df_print: paged
    toc: true
    toc depth: 3
    toc_float: true
    number_sections: true
  md_document:
    variant: markdown_github
    toc: true
    number_sections: true
always_allow_html: true
---
<style type="text/css">
div.main-container {
  max-width: 1800 !important;
}
</style>

```{r setup, include = FALSE}
library(tidyverse)
library(ggplot2)
knitr::opts_chunk$set(echo = T, warning = F, message = F)

prettier_DT <- function(x) DT::datatable(x, options = list(scrollX = TRUE, autoWidth = TRUE), class = "display nowrap")
```

<br>

# Introduction

In this tutorial, we will use [the ENCODE REST API](https://www.encodeproject.org/help/rest-api/) to automate the metadata extraction of experiments related to shRNA knockdown against specific target genes. The functions provided can also be used to extend its functionality (see [appendix 1](#appendix1)).

<br>

# Methods

<br>

## Prerequisites

For this tutorial, we first need the following libraries:

```{r}
library(httr)
library(jsonlite)
library(foreach)
library(tibble)
library(logger)
library(tidyverse)

# ## Logger options
logger::log_threshold(logger::WARN, index = 1)
logger::log_threshold(logger::INFO, index = 2)

logger_layout <- logger::layout_glue_generator(format = '[{time}] [{level}] {msg}')
logger::log_layout(logger_layout, index = 1)
logger::log_layout(logger_layout, index = 2)

source("helperFunctions.R")
```

<br>

## Steps

<br>

### Step 0: Set the extraction parameters

In this step, we need to set the parameters that will determine the program's behaviour. The different options are:

* Valid or required values:
    - **required_cell_lines**: set to `r c("HepG2", "K562")`. It requires each target gene to contain samples for both HepG2 and K562 cell lines.
    - **valid_target_genes**: the specific target genes to focus the search on. Leave as an empty vector `c()` to not filter.
    - **valid_genome annotation**: set to `V29`.
    - **valid_file_format**: set to `bam`.
    - **valid_output_type**: set to `alignments`.
    - **valid_nucleic_acid_type**: set to `polyadenylated mRNA`.
  
* Output files:
    - **output_search**: location to store the results of the ENCODE search. Set to `ENCODE_API_files/all_experiments.tsv`
    - **output_metadata**: location to store the metadata extraction results. Set to `ENCODE_API_files/metadata_samples.tsv`
  
* Input files:
    - **input_target_gene_categories**: path to file extracted from *Van Nostrand et. al.* publication with information about the functional category of the target genes. More information [here](#AddCat).
    - **input_target_gene_NMD**: path to file which contain the list of genes categorized as relevant for nononsense-mediated decay processes.
  
* Other parameters:
    - **overwrite_db**: whether to remove the previous metadata found. Set to `TRUE` if some API call failed during the process so that it only download what the program needs. Set to `F`.
    - **download_method**: it can only be set to "gene_silencing_series" or "experiments". It controls how the script access the API to download the information. Using the gene silencing series, we can access every control and case sample for each cell line with just one call, while using the experiments themselves needs two calls (one for cases and one for controls). It is recommended to set to gene silencing series, unless the ENCODE portal has set the access as restricted for them.

```{r}
## Valid or required values
required_cell_lines <- c("HepG2", "K562")
valid_target_genes <- c() 
valid_genome_annotation <- "V29"
valid_file_format <- "bam"
valid_output_type <- "alignments"
valid_nucleic_acid_type <- "polyadenylated mRNA"

## Other parameters:
overwrite_db <- F
download_method <- "gene_silencing_series" # Other valid option is "experiments"

## Files
main_path <- "Metadata_results_tutorial/"
if(!dir.exists(main_path)) dir.create(main_path)

#### Output files
output_json <- paste0(main_path, "response.json")
output_search <- paste0(main_path, "all_experiments.tsv")
output_metadata <- paste0(main_path, "metadata_samples.tsv")

#### Input files
input_target_gene_categories <- paste0("Additional_files/Target_gene_categories.tsv")
input_target_gene_NMD <- paste0("Additional_files/NMD.txt")

## Additional logger info
log_file <- paste0(main_path, "ENCODE_API.log")
logger::log_appender(logger::appender_file(log_file, append = T), index = 2)
```

<br>

### Step 1: Call the ENCODE API

First, we need to generate an intermediary dataframe containing the information of an ENCODE search. We head to the [ENCODE experiment search portal](https://www.encodeproject.org/search/?type=Experiment), and input the different search filters we are interested in. The following filters are what the program has been tested with (additional search executed in [appendix 1](#appendix1)):

* **Assay**:
    * **Assay title**: shRNA RNA-seq
    * **Target category**: RNA binding protein
    * **Hide control experiments**: yes
* **Biosample**:
    * **Organism**: *Homo sapiens*
* **Quality**:
    * **Status**: released
* **Other filters**:
    * **Data Type**: Experiment

Once we have set the filters, we copy the URL and add `&limit=all&format=json` to return all search results in `json` format. 

```{r}
URL = "https://www.encodeproject.org/search/?type=Experiment&assay_title=shRNA+RNA-seq&target.investigated_as=RNA+binding+protein&control_type!=*&status=released&replicates.library.biosample.donor.organism.scientific_name=Homo+sapiens&limit=all&format=json"
```

Once we have decided on the URL, we execute the `getUrlResponse()` function to call the ENCODE API and convert the returned json object into an `R` list.

```{r}
response_data <- getUrlResponse(URL, output_json)
```

The returned object contains all the information about the search results:

```{r}
names(response_data)
```

<br>

### Step 2: Generate the search summary

Once we have the json object from the ENCODE API, we can extract the relevant information and summarize it in a dataframe. To do so, we execute the function `generateSummary()`, where we specify an output path to store the results in a .tsv file format.

```{r}
summary_df <- generateSummary(response_data, 
                              valid_target_genes = valid_target_genes,
                              output_file = output_search)
```

The summary dataframe contains a target gene and gene silencing series per row, and have the case experiment IDs and cell line for that given target gene. We store the gene silencing series because it contains the information of both the case and control samples, while the experiment IDs only provide information for the case samples. Here is an example of the 10 first elements of the dataframe:

```{r echo = F}
summary_df %>% 
  head(10) %>% 
  `colnames<-`(c("Target gene", "Experiment ID", "Cell line", "Gene silencing series")) %>%
  kableExtra::kbl() %>% 
  kableExtra::kable_classic(full_width = F, html_font = "Cambria")
```

<br>

### Step 3: Retrieval of metadata

In this last step, we loop through every row of the summary dataframe. Since we are using the *Gene silencing series*, only one ENCODE API call for cell line is require, and we can extract all the information of both case and control samples. 

```{r}
metadata_df <- generateMetadata(summary_df, 
                                download_method = download_method,
                                required_cell_lines = required_cell_lines,
                                valid_file_format = valid_file_format,
                                valid_genome_annotation = valid_genome_annotation,
                                valid_output_type = valid_output_type,
                                valid_nucleic_acid_type = valid_nucleic_acid_type,
                                output_file = output_metadata,
                                overwrite_db = overwrite_db)
```

The argument `overwrite_db` of the function determines whether the previous file will be updated or overwritten. If set to `False`, the function will only retrieve the information for the missing target genes, or those which do not have exactly 8 entries. This is because sometimes the API will return bad responses, and we may not want to repeat the process for all target genes.

The other parameters regulate which files or samples we are interested in.

The returned dataframe contains the following columns:

* **target_gene**: the target gene of the shRNA knowckdown.
* **experiment_type**: whether the experiment is case or control for that particular cell line and target gene.
* **cell_line**: the cell type to which the knockdown was executed.
* **gene_silencing_series**: the ENCODE ID of the *Gene Silencing Series* (*GSS*). Information about the *GSS* can be found in the link `https://www.encodeproject.org/gene-silencing-series/Gene_silencing_series/` (i.e. [ENCSR406ZEB](https://www.encodeproject.org/gene-silencing-series/ENCSR406ZEB/))
* **experiment_id**: the ENCODE ID of the experiment in particular. Information about the experiment can be found in the link `https://www.encodeproject.org/experiments/Experiment_ID/` (i.e. [ENCSR047QHX](https://www.encodeproject.org/experiments/ENCSR047QHX/))
* **sample_id**: the ENCODE ID of the sample in particular. Information about the sample can be found in the link `https://www.encodeproject.org/files/Sample_ID/` (i.e. [ENCFF946HGK](https://www.encodeproject.org/files/ENCFF946HGK/))
* **RIN**: the RNA integrity number of the sample. It measures the integrity of the RNA measurements.
* **read_depth**: number of reads per sample.
* **bio_rep**: the isogenic replicate of the sample.
* **tech_rep**: the technical replicate of the sample.
* **sex**: the sex of the sample donor.
* **age**: the age of the sample donor.
* **life_stage**: the life stage of the sample donor.
* **gene_quantification_id**: the ENCODE ID of the associated gene quantification file. It is employed to download the gene expressions (in TPM) before and after the knockdown, to study the efficiency.
* **file_format**: format of the selected sample.
* **output_type**: output type of the selected sample.
* **genome_annotation**: genome annotation of the selected sample.
* **mapped_run_type**: mapped run type of the selected sample.
* **lab**: laboratory where the selected sample was produced.
* **assay**: assay of the selected sample.
* **cellosaurus**: cellosaurus ID of the sample. Usually describes the cell line.
* **biosample**: the ENCODE ID of the biosample from which a case sample was generated.
* **document**: ID of the biosample preparation and characterization document from which the reported knockdown efficiency from ENCODE is extracted.

More information about the metadata provided by ENCODE can be found in their [*Terms and Definitions*](https://www.encodeproject.org/data-standards/terms/) portal.

<br>

### Step 4: Add target gene category {#AddCat}

Additionally, we can add a target gene category based on *Van Nostrand et. al.* [A large-scale binding and functional map of human RNA-binding proteins](https://www.nature.com/articles/s41586-020-2077-3) publication, which adds a category for whether it is relevant for *Splicing regulation*, the *Spliceosome* or an *Exon Junction Complex* (not mutually exclusive). If provided with a list of NMD genes, it will also classify the target genes in this category.

```{r}
metadata_df <- addTargetGeneCategory(metadata_df,
                                     input_target_gene_categories,
                                     input_target_gene_NMD,
                                     output_metadata)
```

The final dataframe is stores in a .tsv file if provided in the `output_file` argument.

<br>

# Results 

The final results can be seen in the following table:

```{r}
prettier_DT(metadata_df)
# metadata_df[1:10, ] %>%
#   kableExtra::kbl() %>%
#   kableExtra::kable_classic(full_width = F, html_font = "Cambria")
```

<br>

## Common metadata

As seen in the table, every row corresponds to a sample related to both a target gene, a cell line and an experiment type. Given our requirements, all rows have a file format of `bam`, an output type of `alignments` and a genome annotation of `V29`. When we study common aspects between the different samples, we observe a clear difference between the two cell lines: **all samples extracted from a same cell line comes from the same donor (sex and age), and from the same tissue (explained in the cellosaurus).**

```{r echo=FALSE}
metadata_df %>% 
  select(cell_line, sex, age, cellosaurus) %>% 
  distinct() %>% 
  kableExtra::kbl() %>%
  kableExtra::kable_classic(full_width = F, html_font = "Cambria")
```

Each cellosaurus corresponds to a different tissue:

* **[CVCL_0004](https://www.cellosaurus.org/CVCL_0004)**: derived from *Pleural effusion*.
* **[CVCL_0027](https://www.cellosaurus.org/CVCL_0027)**: derived from *Liver*.

Other parameters that are constant across the different samples (independently of the cell line) are: the mapped run type, the lab, the assay, the nucleic acid type (required), the extraction method, the fragmentation method, the size selection method and the strand specificity.

```{r echo=FALSE}
metadata_df %>% select(mapped_run_type, lab, assay, nucleic_acid_type, extraction_method, fragmentation_method, size_selection_method, strand_specificity) %>% distinct() %>% prettier_DT()
  # kableExtra::kbl() %>%
  # kableExtra::kable_classic(full_width = F, html_font = "Cambria")#prettier_DT
```

<br>

## Statistics

The total number of target gene founds under our conditions is `r length(unique(metadata_df$target_gene))` (for a total of `r nrow(metadata_df)` samples).

The average RIN is `r metadata_df %>% pull(rin) %>% mean(na.rm = T) %>% round(digits = 2)` [`r min(metadata_df$rin, na.rm = T)`-`r max(metadata_df$rin, na.rm = T)`], considerably high. We do find an experiment without a RIN (experiment ENCSR438UOT, case samples for target gene XPO1 and cell line HepG2), which we define as `NA`.

If we added the target gene categories, we have a total of `r metadata_df %>% select(target_gene, Splicing_regulation, Spliceosome, Exon_junction_complex) %>% distinct() %>% pull(Splicing_regulation) %>% sum(na.rm = T)` marked as *Splicing regulation*, `r metadata_df %>% select(target_gene, Splicing_regulation, Spliceosome, Exon_junction_complex) %>% distinct() %>% pull(Spliceosome) %>% sum(na.rm = T)` marked as *Spliceosome*, `r metadata_df %>% select(target_gene, Splicing_regulation, Spliceosome, Exon_junction_complex) %>% distinct() %>% pull(Exon_junction_complex) %>% sum(na.rm = T)` marked as *Exon junction complex* and `r metadata_df %>% select(target_gene, Splicing_regulation, Spliceosome, Exon_junction_complex, NMD) %>% distinct() %>% pull(NMD) %>% sum(na.rm = T)` marked as *NMD*. If we combine all these categories, a total of `r metadata_df %>% select(target_gene, Splicing_regulation, Spliceosome, Exon_junction_complex, NMD) %>% distinct() %>% filter(if_any(-c(target_gene), ~ . != 0)) %>% nrow()` target genes:

```{r echo=FALSE}
metadata_df %>% select(target_gene, Splicing_regulation, Spliceosome, Exon_junction_complex, NMD) %>% distinct() %>% filter(if_any(-c(target_gene), ~ . != 0)) %>% pull(target_gene)
```

<br>

# Conclusions

With the developed script, it is possible to automate the ENCODE RBP metadata extraction using the [the ENCODE REST API](https://www.encodeproject.org/help/rest-api/). It is also proven that the samples share relevant aspects to allow their direct comparison.

<br>

# Appendix 1: CRISPR metadata download {#appendix1}

Aditionally, the software was tested to download the metadata from CRISPR knockdown. There are a few considerations we would have to keep in mind:

- Some gene silencing series from the CRISPR projects are not public. As such, we have to set the `download_method` to `experiments`, so that the metadata is extracting from the experiment page, and not from the gene silencing series.
- We need to provide a different URL for the search. More precisile, we need to modify the "Assay title" from `shRNA RNA-seq` to `CRISPR RNA-seq`. Other than that, everything else is the same.

With these modifications in mind, the following script will download all metadata found for CRISPR projects where two cell lines are found, each with 4 samples (2 case and 2 control).

```{r eval = F}
## Valid or required values
required_cell_lines <- c("HepG2", "K562")
valid_target_genes <- c() 
valid_genome_annotation <- "V29"
valid_file_format <- "bam"
valid_output_type <- "alignments"
valid_nucleic_acid_type <- "polyadenylated mRNA"

## Other parameters:
overwrite_db <- F
download_method <- "experiments"

## Files
main_path <- "Metadata_results_CRISPR/"
if(!dir.exists(main_path)) dir.create(main_path)

#### Output files
output_json <- paste0(main_path, "response.json")
output_search <- paste0(main_path, "all_experiments.tsv")
output_metadata <- paste0(main_path, "metadata_samples.tsv")

#### Input files
input_target_gene_categories <- paste0("Additional_files/Target_gene_categories.tsv")
input_target_gene_NMD <- paste0("Additional_files/NMD.txt")

## URL to CRISPR experiments
URL = "https://www.encodeproject.org/search/?status=released&type=Experiment&target.investigated_as=RNA+binding+protein&replicates.library.biosample.donor.organism.scientific_name=Homo+sapiens&assay_title=CRISPR+RNA-seq&limit=all&format=json"

## Pipeline
response_data <- getUrlResponse(URL, output_json)
summary_df <- generateSummary(response_data, 
                              valid_target_genes = valid_target_genes,
                              output_file = output_search)
metadata_df <- generateMetadata(summary_df, 
                                download_method = download_method,
                                required_cell_lines = required_cell_lines,
                                valid_file_format = valid_file_format,
                                valid_genome_annotation = valid_genome_annotation,
                                valid_output_type = valid_output_type,
                                valid_nucleic_acid_type = valid_nucleic_acid_type,
                                output_file = output_metadata, 
                                overwrite_db = overwrite_db)
metadata_df <- addTargetGeneCategory(metadata_df,
                                     input_target_gene_categories,
                                     input_target_gene_NMD,
                                     output_metadata)
```

<br>

# Session info

<details>
  <summary>Show/hide</summary>

```{r reproducibility, echo = FALSE}
# Session info
library("sessioninfo")
options(width = 120)
session_info()
```

</details> 