---
title: "Knockdown Efficiency estimation"
subtitle: "TPM from ENCODE"
author: 
- name: "Guillermo Rocamora PÃ©rez"
  affiliation: UCL
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  bookdown::html_document2:
    figure_caption: yes
    code_folding: show
    theme: paper
    highlight: haddock
    df_print: paged
    toc: true
    toc depth: 4
    toc_float: true
    number_sections: true
  md_document:
    variant: markdown_github
    toc: true
    number_sections: true
always_allow_html: true
---
<style type="text/css">
.dataTables_scrollHeadInner{
  width:100% !important;
}
.dataTables_scrollHeadInner table{
  width:100% !important;
}
.code-folding-btn {
  display: none;
}
</style>

```{r setup, include = FALSE}
shhh <- suppressPackageStartupMessages
shhh(library(kableExtra))
shhh(library(ggplot2))
shhh(library(patchwork))
shhh(library(grid))
shhh(library(logger))
shhh(library(biomaRt))
shhh(library(foreach))
shhh(library(doSNOW))
shhh(library(tidyverse))
options(dplyr.summarise.inform = FALSE)

source(here::here("Helper_Functions/hf_KnockdownEfficiency_TPM.R"))

knitr::opts_chunk$set(echo = F, warning = F, message = F, out.width="85%", fig.align = "center", dpi = 300)

prettier_DT <- function(x, order = NULL, dir = "desc"){
  if(is.null(order)){
    DT::datatable(x, options = list(scrollX = TRUE, autoWidth = FALSE, pageLength = 10, 
                                    initComplete = DT::JS("function(){$(this).addClass('compact');}")), 
                  class = "display cell-border compact nowrap")
  }else{
    DT::datatable(x, options = list(scrollX = TRUE, autoWidth = FALSE, pageLength = 10, order = list(order, dir),
                                    initComplete = DT::JS("function(){$(this).addClass('compact');}")), 
                  class = "display cell-border compact nowrap")
  }
}

prettier_kabble_extra <- function(x){
  x[1:10, ] %>%
    kableExtra::kbl() %>%
    kableExtra::kable_classic(full_width = F, html_font = "Cambria")
}

category_labels <- c("Splicing_regulation" = "Splicing regulation", 
                     "Spliceosome"="Spliceosome", 
                     "Exon_junction_complex" = "Exon junction complex",
                     "NMD" = "Nononsense-mediated decay")
```

```{r theme-setup, echo = FALSE, eval = T}
## Custom sciRmdTheme with increased width.
## Please ask guillermorocamora@gmail.com for more details.
sciRmdTheme::set.theme(
  theme = "default",
  color = NULL,
  header.sticky = FALSE,
  list.group.icon = "arrow",
  font.family = "Arial",
  font.color = "black",
  header.color = "darkblue"
)

custom_gg_theme <- theme(plot.title = element_text(size = 12, face = "bold"),
                  panel.border = ggplot2::element_rect(colour = "black", fill = NA, linewidth = 1),
                  axis.text.x = ggplot2::element_text(color = "black", size = 8, angle = 0, hjust = 0.5),
                  axis.text.y = ggplot2::element_text(color = "black", size = 8),
                  axis.title.x = ggplot2::element_text(face = "bold", size = 11, margin=margin(5,0,0,0)),
                  axis.title.y = ggplot2::element_text(face = "bold", size = 11, margin=margin(0,10,0,0)),
                  panel.grid.minor = element_line(color = "#444444", linewidth  = 0.05, linetype = 2),
                  panel.grid.major.y = element_line(color = "#444444", linewidth  = 0.05, linetype = 2),
                  panel.grid.major.x = element_line(color = "#444444", linewidth  = 0.1),
                  panel.background = element_rect(fill = "#FBFBFB"),
                  legend.title = element_text(size=12),
                  legend.text = element_text(size=10),
                  legend.position = "top",
                  legend.key = element_rect(color="black"),
                  legend.key.size = unit(1, 'lines'),
                  strip.text.x = element_text(color = "black", face = "bold", size = 9),
                  strip.background = element_rect(color="black", linewidth=1, linetype="solid"),
                  strip.text.y = element_text(color = "black", face = "bold", size = 9),
                  plot.margin = margin(0.5, 0.5, 0.2, 0.5, "cm"))
```
 
# Objective

In this tutorial, we aim to estimate the knockdown efficiency of the shRNA knockdown targeted genes by comparing the TPM values between control and cases. This method is not the recommended approach to measure the knockdown effiency, please refer to the Western Blotting method extracted from ENCODE.

# Methods

First, we needed a list of target RBPs to study. As with the previous studies, we will employ the list of 356 RBPs published by *Van Nostrand et. al.* [A large-scale binding and functional map of human RNA-binding proteins](https://www.nature.com/articles/s41586-020-2077-3) categorized as splicing regulation, spliceosome or exon-junction complex. We also added a list of 118 genes classified as involved in nononsense-mediated decay processes. In total, 56 of those genes were found as ENCODE projects with the same number of experiments and similar metadata.

Once we have set the list of ENCODE projects to focus our studies, we need to download the gene expressions for all samples. We have to consider:

- Each sample studied in the ENCODE RBP analysis has an associated `gene_quantification.tsv` file to download.
- The files are the byproduct of using the software *RSEM* over both `FASTQ` files. In all ENCODE projects, we find two sets of `FASTQ` files which contain the first and second mates of the paired-end reads. Since we have 2 isogenic replicates in all experiments, a total of 4 `FASTQ` files are processed. To produce these files, one can follow the *RMSE* [tutorial](https://www.encodeproject.org/experiments/ENCSR896CFV/) and the ENCODE [pipeline information](https://www.encodeproject.org/documents/6354169f-86f6-4b59-8322-141005ea44eb/@@download/attachment/Long%20RNA-seq%20pipeline%20overview.pdf).
- Therefore, a total of 4 files for cases and 4 files for controls are downloaded. Each with 2 for K562 cell line and 2 for hepG2 cell line.
- Out of the 56 initial target genes, only 51 have their gene expressions available to download.
  
Then, we need to extract the TPM value of the target gene in that particular experiment.

- Since the `gene_quantification.tsv` files use the gene ENSEMBL notation, we use the `biomaRt` package to translate from the HGNC nomenclature to their ENSEMBL Id.
- The provided tables in `gene_quantification` contain many useful columns: TPM, posterior mean estimates of the TPM (using Gibbs sampling), lower and upper CI bounds and TPM coefficient of quartile variation. We opted to use the TPM column as is. 
    
Once we have the TPM values before and after the knockdown, we evaluate the knockdown efficiency as defined in equation (1). For that, we first need to calculate the average TPM values for each cluster (case/control), independent of the cell line. A knockdown efficiency of 100% implies that the target gene was not found in the case sample, while a 0% indicates that no change in TPM was observed.

$$
kEff\;\% =\left(1-\frac{TPM_{avg,\;case}}{TPM_{avg,\;control}}\right)*100\%\tag{1}
$$

# Steps

## Step 0: Load previous metadata and set script parameters {-}

First, we need to indicate the file from which to obtain the metadata of the samples. In our case, it is the path to the file generated from the metadata extraction tutorial. We also need to specify certain parameters to the algorithm, like the directory to save the `gene_quantification.tsv` files, or the number of cores to use.

```{r, echo = T}
## Input Files
main_metadata_path <- here::here("Metadata_Results/")
metadata_path <- paste0(main_metadata_path, "metadata_samples.tsv")
metadata_TPM_output <- paste0(main_metadata_path, "metadata_TPM_kEff.tsv")

## Define the algorithm variables
main_path = here::here("Knockdown_Efficiency/TPM/RBPs/")
download_only = F
download_cores = 16
overwrite_results = F

## Generate the variables ----
metadata <- readr::read_delim(metadata_path, show_col_types = F) %>% as_tibble()

target_RBPs <- metadata %>%
  dplyr::filter(if_any(c(Splicing_regulation, Spliceosome, Exon_junction_complex, NMD), ~ . != 0)) %>%
  dplyr::filter(!is.na(gene_quantification_id)) %>%
  dplyr::pull(target_gene) %>%
  unique()

## Convert HGCN nomenclature to ENSEMBL ID
ensembl_target_RBPs <- translateGenes(target_RBPs)

metadata_filtered <- metadata %>%
  dplyr::filter(target_gene %in% target_RBPs, !is.na(gene_quantification_id)) %>%
  dplyr::select(target_gene, cell_line, experiment_type, experiment_id, gene_quantification_id) %>%
  dplyr::mutate(path = paste0(main_path, target_gene, "/", experiment_type, "/")) %>%
  dplyr::left_join(ensembl_target_RBPs, by = c("target_gene" = "hgnc_symbol"), multiple = "all") %>%
  dplyr::relocate(ensembl_gene_id, .before = cell_line)
```

## Step 1: Download the `gene_quantification` files {-}

Next, we use the functions `createDirectories()` and `downloadGeneQuantifications()` to generate the folder structure where the gene quantifications will be stored and to download them from the ENCODE portal. The download process can be parallelized, since most of the download times is spent stablishing the connection to the server and not downloading the file itself (meaning that it is unlikely to saturate the download bandwidth).

```{r, echo = T, message=F}
# Create the directories
createDirectories(target_RBPs, metadata_filtered)

# Download the files and add a column with their path
metadata_quantifications <- downloadGeneQuantifications(metadata_filtered,
                                                        download_cores, 
                                                        overwrite_results,
                                                        silent = T)
```

## Step 2: TPM extraction {-}

Using the package `biomaRt`, we converted the HGCN nomenclature of the genes into their ENSEMBL ID. Then, we use the function `extractTPM()` to read the gene quantification file and extract the TPM for the target gene of that particular samples.

```{r, echo = F, eval = T}
file_name <- here::here("metadata_TPM.rds")
if(file.exists(file_name)){
  metadata_TPM <- readRDS(file_name)
}else{
  metadata_TPM <- extractTPM(metadata_quantifications)
}
```

```{r, echo = T, eval = F}
# Extract the TPMs
metadata_TPM <- extractTPM(metadata_quantifications)
```

Results are shown in the following table:

```{r, echo = F}
metadata_TPM %>% 
  dplyr::select(-path, -experiment_id, -file_path) %>%
  `colnames<-`(c("Target gene", "ENSEMBL ID", "Cell line", "Experiment type", "Gene quantification ID", "TPM")) %>%
  .[sample(1:nrow(.), 10), ] %>%
  kableExtra::kbl(booktabs = T, linesep = "") %>%
  kableExtra::kable_classic(full_width = T, "hover", "striped", html_font = "Cambria", font_size = 14)
```

## Step 3: Estimating the Knockdown Efficiency

Lastly, we use the function `generateKnockdownEfficiency()` to apply the knockdown efficiency formula. Results are obtained for both all samples at the same time and divided by cell lines. In most cases, the overall knockdown efficiency is the average of the efficiencies for each cell line (with some notable exceptions).

```{r, echo = T}
# Generate and save the knockdown efficiencies
metadata_kEff <- generateKnockdownEfficiency(metadata_TPM,
                                             output_file = metadata_TPM_output)
```

Final results are stored in disk:

```{r, echo = F}
metadata_kEff %>%
  dplyr::select(-Avg_TPM_case, -Avg_TPM_control) %>%
  `colnames<-`(c("Target gene", "kEff [%]", "kEff HepG2 [%]", "kEff K562 [%]")) %>%
  dplyr::mutate(across(where(is.numeric), round, 1)) %>%
  .[sample(1:nrow(.), 10), ] %>%
  kableExtra::kbl(booktabs = T, linesep = "") %>%
  kableExtra::kable_classic(full_width = F, "hover", "striped", html_font = "Cambria", font_size = 14)
```

# Results

## Knockdown Efficiency

The average knockdown efficiency is `r metadata_kEff$kEff %>% mean %>% round`%, ranging from `r metadata_kEff$kEff %>% min %>% round` to `r metadata_kEff$kEff %>% max %>% round`% in some cases. In the following visualization, we represent in the X-axis the estimated knockdown efficiency for every studied target RBP/NMD, split by their functional category:

```{r app1-kEff-category-graph, echo = F, fig.height=9, fig.width=7.2, dpi = 300}
metadata_RBPs <- metadata %>% 
  dplyr::filter(target_gene %in% target_RBPs) %>%
  tidyr::pivot_longer(c("Splicing_regulation", "Spliceosome", "Exon_junction_complex", "NMD"), names_to = "Category") %>%
  dplyr::filter(value == 1) %>%
  dplyr::select(-value) %>%
  dplyr::distinct(target_gene, sample_id, .keep_all = T)

kEff_category <- metadata_kEff %>%
  dplyr::left_join(metadata_RBPs %>% select(target_gene, Category) %>% dplyr::distinct(),
                   by = "target_gene") %>%
  dplyr::arrange(-kEff) %>%
  dplyr::mutate(target_gene = factor(target_gene, levels = .$target_gene),
                Category = factor(Category, levels = c("Splicing_regulation", "Spliceosome", "NMD", "Exon_junction_complex")))

ggplot(kEff_category) +
  geom_bar(aes(x = target_gene, y = kEff, fill = Category), stat = "identity", color = "black", linewidth = 0.25, width = 0.8) +
  viridis::scale_fill_viridis(begin = 0.2, discrete = T, labels = category_labels, guide = guide_legend(nrow=2, byrow=TRUE)) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.02))) +
  scale_x_discrete(expand = expansion(add = c(0.7, 0.7))) +
  coord_flip(ylim = c(0, 100)) +
  labs(x = "Target shRNA knockdown gene", y = "Knockdown Efficiency [%]") +
  ggforce::facet_col(vars(Category),
                     scales = "free_y",
                     space = "free",
                     labeller = labeller(Category = category_labels)) +
  custom_gg_theme + theme(plot.title = element_text(size = 20, face = "bold"),
                         legend.title = element_text(size=10),
                         legend.text = element_text(size=8),
                         legend.key.size = unit(0.75, 'lines'))

#ggsave(file = "images/kEff_category.png", width = 183, height = 183*9/7.2, units = "mm", dpi = 300)
```

## Cell line influence

If we study with further details the table provided with the results, we can observe that, in some instances, there are RBPs in which the knockdown efficiency is significantly different between the two different cell lines. 

Let's study the gene `MATR3` for example, where we observe that the $TPM_{avg,\, control}$ of <span style="background-color: #c2ffa1">K562</span> is around 332 TPM while for <span style="background-color: #ffaca1">HepG2</span> is only 152 TPM.

```{r app1-tpm-MATR3, echo = F}
metadata_TPM %>%
  dplyr::filter(target_gene == "MATR3") %>%
  dplyr::select(target_gene, cell_line, experiment_type, TPM, gene_quantification_id) %>%
  dplyr::arrange(experiment_type, cell_line) %>%
  `colnames<-`(c("Target gene", "Cell line", "Cluster", "TPM", "Gene quantification ID")) %>%
  kableExtra::kbl(booktabs = T, linesep = "") %>%
  kableExtra::kable_classic(full_width = F, "hover", "striped", html_font = "Cambria", font_size = 14) %>%
  kableExtra::row_spec(5:6, color = "black", background = "#ffaca1") %>%
  kableExtra::row_spec(7:8, color = "black", background = "#c2ffa1")
```

If we study the knockdown efficiency for each cell line individually, we observe a $kEff$ of 66.8% for HepG2 and 79% for K562:

```{r app1-kEff-MATR3, echo = F}
metadata_TPM %>%
   dplyr::filter(target_gene == "MATR3") %>% 
   dplyr::select(target_gene, cell_line, experiment_type, TPM, gene_quantification_id) %>%
   dplyr::group_by(target_gene, cell_line, experiment_type) %>%
   dplyr::summarize(avg_TPM = mean(TPM)) %>%
   tidyr::pivot_wider(c(target_gene, cell_line), names_from = "experiment_type", values_from = avg_TPM) %>%
   dplyr::mutate(Diff = (1-case/control)*100) %>%
   dplyr::mutate(Diff = paste0(round(Diff, 1), "%")) %>%
   `colnames<-`(c("Target gene", "Cell line", "TPM Cases", "TPM Controls", "kEff")) %>%
   kableExtra::kbl(booktabs = T, linesep = "") %>%
   kableExtra::kable_classic(full_width = F, "hover", "striped", html_font = "Cambria", font_size = 14)
```

In some cases, the difference in $kEff$ between the two cell lines is significantly greater. In the following table and graph, we show the $kEff$ for each cell line and the difference between the two:

```{r app1-kEff-cell_line, echo = F}
kEff_cell_line <- metadata_TPM %>%
  dplyr::select(target_gene, cell_line, experiment_type, TPM, gene_quantification_id) %>%
  dplyr::group_by(target_gene, cell_line, experiment_type) %>%
  dplyr::summarize(avg_TPM = mean(TPM)) %>%
  tidyr::pivot_wider(c(target_gene, cell_line), names_from = "experiment_type", values_from = avg_TPM) %>%
  dplyr::mutate(Diff = (1-case/control)*100)

kEff_cell_line %>%
  dplyr::group_by(target_gene) %>%
  tidyr::pivot_wider(id_cols = target_gene, names_from = cell_line, values_from = Diff) %>%
  dplyr::mutate(Diff = abs(HepG2 - K562)) %>%
  `colnames<-`(c("Target gene", "HepG2 kEff %", "K562 kEff %", "Difference abs(HepG2 - K562)")) %>%
  dplyr::mutate(across(where(is.numeric), round, 1)) %>%
  DT::datatable(rownames = F,
                options = list(scrollX = TRUE,
                               autoWidth = FALSE,
                               pageLength = 10,
                               order = list(3, "desc")),
                class = "display cell-border compact nowrap")
```

```{r app1-kEff-cell_line-graph, echo = F, fig.height=12, fig.width=7.2, dpi = 300}
test <- kEff_cell_line %>%
  dplyr::group_by(target_gene) %>%
  tidyr::pivot_wider(target_gene, names_from = cell_line, values_from = Diff) %>%
  dplyr::mutate(Diff = abs(HepG2 - K562)) %>%
  dplyr::left_join(metadata_RBPs %>% select(target_gene, Category) %>% distinct(),
                   by = "target_gene") %>%
  dplyr::arrange(-HepG2) %>%
  dplyr::mutate(target_gene = factor(target_gene, levels = .$target_gene),
                Category = factor(Category, levels = c("Splicing_regulation", "Spliceosome", "NMD", "Exon_junction_complex")),
                type = ifelse(HepG2 > K562, "HepG2", "K562"))

ggplot(test) +
  geom_segment(aes(x = target_gene, y = HepG2, yend = K562, xend = target_gene, col = "Segment_col"), 
               linewidth = 1.5) +
  scale_color_manual(name = "Difference", labels = "", values = c("Segment_col" = "#5dc863"),
                     guide = guide_legend(override.aes = list(width = 10), keywidth = 2)) +
  ggnewscale::new_scale_color() + 
  geom_point(aes(x = target_gene, y = HepG2, color = "HepG2"), size = 1.5) + 
  geom_point(aes(x = target_gene, y = K562, color = "K562"), size = 1.5) + 
  scale_color_manual(name = "Cell line",
                    labels = c("HepG2", "K562"),
                    values = c("HepG2" = "black", "K562" = "#2181ff"),
                    guide = guide_legend(order = 1)) + 
  coord_flip(ylim = c(0, 100)) +
  ggforce::facet_col(vars(Category), 
                     scales = "free_y", space = "free", 
                     labeller = labeller(Category = category_labels)) +
  labs(x = "Target gene", y = "Knockdown efficiency [%]") + ggtitle("Knockdown Efficiency by cell line") + 
  scale_y_continuous(expand = expansion(mult = c(0, 0))) + 
  custom_gg_theme + theme(legend.key = element_blank(),
                          strip.text.x = element_text(color = "black", face = "bold", size = 11),
                          plot.title = element_text(size = 16),
                          legend.margin = margin(0, 1.5, 0, 0, unit="cm"))
#ggsave(file = "images/kEff_category_cell_line.png", width = 183, height = 183*9/7.2, units = "mm", dpi = 300)
```

We represent in the X and Y-axis the same information as before, but divided by the cell line (black for HepG2 and blue for K562). The green line represents the distance between the two measurements.

For our analysis, we grouped both cell lines independently without any considerations. The previous graph shows that the knockdown effects are not always consistent in both cell lines. These results could have a relevant effect in the mis-splicing ratio of introns found in the more affected genes (e.g. U2AF2, SRSF1).

<!-- For example, for target gene U2AF2, the variation in TPM from <span style="background-color: #c2ffa1">control</span> to <span style="background-color: #ffaca1">cases</span> in the HepG2 cell line produces a knockdown efficiency of 1.7%, while nearly a 70% efficiency for the K562 cell line. -->

<!-- ```{r app1-kEff-U2AF2, echo = F} -->
<!-- metadata_TPM %>% -->
<!--   dplyr::filter(target_gene == "U2AF2") %>% -->
<!--   dplyr::select(target_gene, cell_line, experiment_type, TPM, gene_quantification_id) %>% -->
<!--   dplyr::arrange(cell_line, experiment_type) %>% -->
<!--   `colnames<-`(c("Target gene", "Cell line", "Cluster", "TPM", "Gene quantification ID")) %>% -->
<!--   kableExtra::kbl(booktabs = T, linesep = "") %>% -->
<!--   kableExtra::kable_classic(full_width = F, "hover", "striped", html_font = "Cambria", font_size = 14) %>% -->
<!--   kableExtra::row_spec(1:2, color = "black", background = "#ffaca1") %>% -->
<!--   kableExtra::row_spec(3:4, color = "black", background = "#c2ffa1") -->
<!-- ``` -->
