---
title: "WB/qt-PCR Knockdown efficiency extraction"
author: 
- name: "Guillermo Rocamora PÃ©rez"
  affiliation: UCL
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  bookdown::html_document2:
    figure_caption: yes
    code_folding: show
    theme: paper
    highlight: haddock
    df_print: paged
    toc: true
    toc depth: 4
    toc_float: true
    number_sections: true
  md_document:
    variant: markdown_github
    toc: true
    number_sections: true
always_allow_html: true
---
<style type="text/css">
.dataTables_scrollHeadInner{
  width:100% !important;
}
.dataTables_scrollHeadInner table{
  width:100% !important;
}
.code-folding-btn {
  display: none;
}
</style>

```{r setup, include = FALSE}
shhh <- suppressPackageStartupMessages
shhh(library(kableExtra))
shhh(library(ggplot2))
shhh(library(patchwork))
shhh(library(grid))
shhh(library(logger))
shhh(library(biomaRt))
shhh(library(foreach))
shhh(library(doSNOW))
shhh(library(tidyverse))
options(dplyr.summarise.inform = FALSE)

source(here::here("Helper_Functions/hf_KnockdownEfficiency_WB.R"))

knitr::opts_chunk$set(echo = F, warning = F, message = F, out.width="85%", fig.align = "center", dpi = 300)

prettier_kabble_extra <- function(x){
  x[1:10, ] %>%
    kableExtra::kbl() %>%
    kableExtra::kable_classic(full_width = F, html_font = "Cambria")
}

prettier_DT <- function(x, order = NULL, dir = "desc"){
  if(is.null(order)){
    DT::datatable(x, options = list(scrollX = TRUE, 
                                    autoWidth = FALSE, 
                                    pageLength = 10), 
                  class = "display cell-border compact nowrap")
  }else{
    DT::datatable(x, options = list(scrollX = TRUE, 
                                    autoWidth = FALSE, 
                                    pageLength = 10, 
                                    order = list(order, dir)), 
                  class = "display cell-border compact nowrap")
  }
}

category_labels <- c("Splicing_regulation" = "Splicing regulation", 
                     "Spliceosome"="Spliceosome", 
                     "Exon_junction_complex" = "Exon junction complex",
                     "NMD" = "Nononsense-mediated decay")
```

```{r theme-setup, echo = FALSE, eval = T}
sciRmdTheme::set.theme(
  theme = "default",
  color = NULL,
  header.sticky = FALSE,
  list.group.icon = "arrow",
  font.family = "Arial",
  font.color = "black",
  header.color = "darkblue"
)

custom_gg_theme <- theme(
  plot.title = element_text(size = 12, face = "bold"),
  panel.border = ggplot2::element_rect(colour = "black", fill = NA, linewidth = 1),
  axis.text.x = ggplot2::element_text(color = "black", size = 8, angle = 0, hjust = 0.5),
  axis.text.y = ggplot2::element_text(color = "black", size = 8),
  axis.title.x = ggplot2::element_text(face = "bold", size = 11, margin = margin(5, 0, 0, 0)),
  axis.title.y = ggplot2::element_text(face = "bold", size = 11, margin = margin(0, 10, 0, 0)),
  panel.grid.minor = element_line(color = "#444444", linewidth = 0.05, linetype = 2),
  panel.grid.major.y = element_line(color = "#444444", linewidth = 0.05, linetype = 2),
  panel.grid.major.x = element_line(color = "#444444", linewidth = 0.1),
  panel.background = element_rect(fill = "#FBFBFB"),
  legend.title = element_text(size = 12),
  legend.text = element_text(size = 10),
  legend.position = "top",
  legend.key = element_rect(color = "black"),
  legend.key.size = unit(1, "lines"),
  strip.text.x = element_text(color = "black", face = "bold", size = 9),
  strip.background = element_rect(color = "black", linewidth = 1, linetype = "solid"),
  strip.text.y = element_text(color = "black", face = "bold", size = 9),
  plot.margin = margin(0.5, 0.5, 0.2, 0.5, "cm")
)
```

# Objective

In this tutorial, we will focus on downloading and extracting the reported knockdown efficiency in the ENCODE shRNA knockdown experiments.

# Methods

First, we needed a list of target RBPs to study. As with the previous studies, we will employ the list of 356 RBPs published by Nostrand et al. which has been categorized as splicing regulation, spliceosome or exon-junction complex. We also added a list of 118 genes classified as involved in nononsense-mediated decay processes. In total, 56 of those genes were found as ENCODE projects with the same number of experiments and similar metadata.

For most experiments (being an experiment a collection of 2 isogenic replicates for one cell line and experiment type), the knockdown efficiency is reported in their *Graveley Lab shRNA knockdown followed by RNA-seq Biosample Preparation and Characterization Document* (see [example](https://www.encodeproject.org/documents/e5391dd5-cf87-451c-97e0-fe9967170285/@@download/attachment/U2AF2-LV08-64.pdf)). They use two alternative techniques to evaluate the efficiency: qRT-PCR and Western blotting.

In both situations, they measure the depletion level in the shRNA treated sample against a control non-target shRNA sample. The depletion is normalized using the gene GAPDH as a control. The technique we are most interested in is the Western blotting, a powerful and important procedure to analyse the detection of proteins, particularly in low abundance [@KURIEN2006283]. As such, it represents the actual concentration of protein in the sample and not the RNA abundance.

The main procedure to obtain the reported knockdown efficiency for a particular experiment is the following:

1. Identify if the report is available for the particular experiment. See for example the [ADAR K562 experiment](https://www.encodeproject.org/experiments/ENCSR164TLB/). At the bottom of the page, we need to find any document with the description: *biosample protocol for shRNA followed by RNA-seq*. The procedure is executed in the ENCODE metadata extraction framework, using [the ENCODE REST API](https://www.encodeproject.org/help/rest-api/). If the document is found, we store the information in the metadata from the samples.

2. Download the documents for every interest RBP/NMD experiment. Even if the document is found, there are many inconsistencies inside the ENCODE experiments that make it hard to identify the download link (e.g. [some](https://www.encodeproject.org/experiments/ENCSR220TBR/) documents has the suffix `_update`, [others](https://www.encodeproject.org/experiments/ENCSR605MFS/) have an alternative HGCN name for the gene). Because of that, not all possible reports are guaranteed to be found.

3. Since the knockdown efficiency is written in an image inside the pdf, we employed the `extract -images` [script](https://pymupdf.readthedocs.io/en/latest/module.html#extracting-fonts-and-images) from the *fitz* module within the python package *PyMuPDF* version 1.21.1. There are several figures inside the pdf, but fortunately the Western blotting results are always the last figure. In figure \@ref(fig:wbimages) we can see an example of the figures that contain the knockdown efficiency inside the reports.

```{r wbimages, echo=FALSE, out.width="49%", out.height="20%", fig.cap="Different images in which the knockdown efficiency is reported.", fig.show='hold', fig.align='center'}
knitr::include_graphics(c("images/ENCLB155EFP_Western_Blot_Analysis.png","images/ENCLB867RJA_Western_Blot_Analysis.png"))
``` 

4. Once the images are extracted, we need to employ an OCR (Optical Character Recognition) algorithm to identify the text within the figures. To do so, we employed the open-source [Tesseract-OCR](https://github.com/tesseract-ocr/tesseract#tesseract-ocr) engine currently developed by Google. Its main advantages are the simplicity of use, overall efficiency and easy implementation in a python pipeline using the package [pytesseract](https://pypi.org/project/pytesseract/) version 0.3.10. Using the R library [reticulate](https://rstudio.github.io/reticulate/) version 1.26, we called the python functions to extract the text from the image. Some considerations about the procedure:

    - The images contain a lot of other information and text not related to the Western blot results. This hinders the identification of the knockdown efficiencies from the output of the OCR algorithm. Thus, we decided to crop the image to only show the percentages and the method to measure the efficiency. See figure \@ref(fig:croppedimage).
    - The images extracted from the ENCODE report are usually in high resolution. However, the tesseract-ocr training probably used lower resolution images. Therefore, the accuracy was quite low in our initial tests. To solve the issue we had to resize the images to a lower resolution. We found that the most accurate resolution was 25% of the original.
    - Because of the large amount of total reports from ENCODE, the accuracy could not be manually estimated. As such, a 100% accuracy is not guaranteed.


```{r croppedimage, echo=FALSE, out.width="49%", out.height="20%", fig.cap="Cropped image from which we extract the text.", fig.show='hold', fig.align='center'}
knitr::include_graphics("images/cropped_image.png")
```

5. The last step was to summarize the results. Since two knockdown efficiencies were provided for each experiment, we estimated that the knockdown efficiency for that particular cell line was the average of the two. In the data-analysis pipeline we made no distinction between the two cell lines, thus the final knockdown efficiency was estimated as the average of the two cell lines.

# Steps

## Prerequisites

For this tutorial, we need to install the [Tesseract-OCR](https://github.com/tesseract-ocr/tesseract#tesseract-ocr) engine. We also need the library [reticulate](https://rstudio.github.io/reticulate/) version 1.26 to use the python modules [Pillow](https://pillow.readthedocs.io/en/stable/) and [pytesseract](https://pypi.org/project/pytesseract/). 

Therefore, we need to load the necessary libraries and helper functions:

```{r}
# Required libraries
library(logger)
library(foreach)
library(doSNOW)
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(reticulate))

# Load additional helper functions
source("helperFunctions.R")

# Load python interface
use_virtualenv("UCL")
pytesseract <- import("pytesseract")
PIL <- import("PIL")
```

## Step 1: Load previous metadata and set script parameters {-}

First, we specify where the input metadata is located and where the output files will be stored. We also specify some parameters for the scripts, like the path to store the downloaded files, the number of download cores to use, the resize percentage to apply to the figures and the minimum width to apply the resize (more information on that later).

```{r}
# Files
metadata_path <- "../Metadata_results/metadata_samples.tsv"
metadata_WB_output <- "../Metadata_results/metadata_WB_kEff.tsv"
metadata_PCR_output <- "../Metadata_results/metadata_PCR_kEff.tsv"
metadata <- readr::read_delim(metadata_path, show_col_types = F) %>% as_tibble()

# Define the algorithm variables
main_path = "RBPs/"
download_cores = 16
overwrite_results = F
resize_perc = 0.25
min_width = 600

# Generate the variables
target_RBPs <- metadata %>%
  dplyr::filter(if_any(c(Splicing_regulation, Spliceosome, Exon_junction_complex, NMD), ~ . != 0)) %>%
  dplyr::filter(!is.na(document)) %>%
  dplyr::pull(target_gene) %>%
  unique()

metadata_RBPs <- metadata %>% 
  filter(target_gene %in% target_RBPs) %>%
  pivot_longer(c("Splicing_regulation", "Spliceosome", "Exon_junction_complex", "NMD"), names_to = "Category") %>%
  filter(value == 1) %>%
  select(-value) %>%
  distinct(target_gene, sample_id, .keep_all = T)

metadata_filtered <- metadata %>%
  dplyr::filter(target_gene %in% target_RBPs, !is.na(document)) %>%
  dplyr::select(target_gene, cell_line, experiment_id, biosample, bio_rep, document, biosample_alias) %>%
  dplyr::mutate(path = paste0(main_path, target_gene, "/", experiment_id, "/", bio_rep, "/"))
```

## Step 2: Download the biosample characterization documents {-}

Next step is to generate the subfolder structures where the biosample characterization files will be stored (by default `main_path/target_gene/experiment_id/biosample_replicate/`). We use the function `downloadCharacterizationDocuments()` to download the documents into the specified paths. 

```{r}
## Create the directories
createDirectories(target_RBPs, metadata_filtered)

## Download the files and add a column with their path
metadata_documents <- downloadCharacterizationDocuments(metadata_filtered, 
                                                        download_cores, 
                                                        overwrite_results,
                                                        silent = T)
```

It is recommended to test for the existence of the documents, since it is possible that some experiments have their documents file names modified, or they might not be available to download.

```{r}
# Check the existence of the files
for(row_index in seq(nrow(metadata_documents))){
  row = metadata_documents[row_index, ]
  file_path <- row$file_path
  experiment_id <- row$experiment_id
  path <- row$path
  
  if(!file.exists(file_path) || file.info(file_path)$size < 10) cat(paste0("Error for row ", row_index, " experiment ", experiment_id, "! File path ", path, "\n"))
}
```

## Step 3: Extract the images from the documents {-}

Once we have all characterization documents available, we use the python package [PyMuPDF](https://pymupdf.readthedocs.io/en/latest/) version 1.21.1. The function `extractImages()` extract the images and keeps only the last one, which contains the reported efficiency for both methods.

```{r}
# Extract the images of all files
metadata_images <- extractImages(metadata_documents, 
                                 overwrite_results = overwrite_results)
```

## Step 4: Extract the text from the images {-}

The next step is to process the images from the step above and to extract the written text in them to obtain the reported knockdown efficiencies. As mentioned, we use the python modules *pytesseract* and *Pillow*. First, we crop the image and reduce its dimensions (tesseract-OCR works better with lower resolution images). We apply the basic tesseract-OCR configuration to the image and simply modify the returned string to obtain the results in data.frame format.

Since there are two types of images, we need to consider both scenarios, differenciated by the number of columns that tesseract-OCR reads from the image. The efficiencies are extracted and combined per cell line (i.e. two efficiencies are provided per cell line, but the average is reported). If no valid efficiency is extracted from the image, an `NA` is returned instead:

```{r}
# Extract text from images. This cannot be separated into an external function.
# Probably because of some incompatibility with the reticulate library to use
# python.
metadata_kEff <- foreach(row_index = seq(nrow(metadata_images)), .combine = dplyr::bind_rows) %do%{
  row = metadata_images[row_index, ]
  path <- row$path
  image_path <- row$image_path
  
  ## If the image does not exists, return the unmodified row
  if(!file.exists(image_path)) return(row)
  
  image <- PIL$Image$open(image_path)
  image_cropped <- image$crop(list(0, image$height*0.9, 0.8*image$width, image$height))
  image_small <- resizeImage(image, resize_perc, min_width)
  
  image_small$save(paste0(path, "cropped_image.png"))
  text_df <- pytesseract$image_to_string(image_small) %>%
    str_replace_all("\\f", "") %>%
    str_split("\\n", simplify = T) %>%
    str_split(" ", simplify = T) %>%
    .[1:2, ] %>%
    as_tibble()
  
  if(ncol(text_df) == 5){
    kEff_df <- text_df %>% 
      `colnames<-`(c("method", "K562_1", "K562_2", "HepG2_1", "HepG2_2")) %>%
      dplyr::mutate(across(-method, function(x) as.numeric(sub("%", "", x)))) %>%
      dplyr::mutate(K562 = rowMeans(dplyr::select(., K562_1, K562_2)),
                    HepG2 = rowMeans(dplyr::select(., HepG2_1, HepG2_2))) %>%
      dplyr::select(method, K562, HepG2)
    
    row$WB_HepG2 <- kEff_df %>% dplyr::filter(method == "Western") %>% dplyr::pull(HepG2)
    row$WB_K562 <- kEff_df %>% dplyr::filter(method == "Western") %>% dplyr::pull(K562)
    
    row$PCR_HepG2 <- kEff_df %>% dplyr::filter(method != "Western") %>% dplyr::pull(HepG2)
    row$PCR_K562 <- kEff_df %>% dplyr::filter(method != "Western") %>% dplyr::pull(K562)
  }else if(ncol(text_df) == 3){
    kEff_df <- text_df %>% 
      `colnames<-`(c("method", "cell_line_1", "cell_line_2")) %>%
      dplyr::mutate(across(-method, function(x) as.numeric(sub("%", "", x)))) %>%
      dplyr::mutate(cell_line = rowMeans(dplyr::select(., cell_line_1, cell_line_2))) %>%
      dplyr::select(method, cell_line)
    
    row$WB_HepG2 <- ifelse(row$cell_line == "HepG2", kEff_df %>% dplyr::filter(method == "Western") %>% dplyr::pull(cell_line), NA)
    row$WB_K562 <- ifelse(row$cell_line == "K562", kEff_df %>% dplyr::filter(method == "Western") %>% dplyr::pull(cell_line), NA)
    
    row$PCR_HepG2 <- ifelse(row$cell_line == "HepG2", kEff_df %>% dplyr::filter(method != "Western") %>% dplyr::pull(cell_line), NA)
    row$PCR_K562 <- ifelse(row$cell_line == "K562", kEff_df %>% dplyr::filter(method != "Western") %>% dplyr::pull(cell_line), NA)
  }else{
    print(paste0("Error in row ", row_index, ". Columns are not valid"))
    row$WB_HepG2 <- NA
    row$WB_K562 <- NA
    row$PCR_HepG2 <- NA
    row$PCR_K562 <- NA
  }
  
  return(row)
}
```

An important step is to test for consistency between the returned efficiencies. In many experiments, each sample's document reports the efficiency for the whole target gene. As such, the same image is processed up to 4 times per target gene. It is important to test that all iterations of the same image returns the same output:

```{r}
# Test for consistency between the cell lines
for(target_RBP in target_RBPs){
  metadata_RBP <- metadata_kEff %>%  filter(target_gene == target_RBP)
  WB_HepG2 <- metadata_RBP$WB_HepG2
  WB_K562 <- metadata_RBP$WB_K562
  PCR_HepG2 <- metadata_RBP$PCR_HepG2
  PCR_K562 <- metadata_RBP$PCR_K562
  
  if(length(unique(na.omit(WB_HepG2))) > 1) print(paste0("Error in RBP ", target_RBP, " cell line HepG2, method WB."))
  if(length(unique(na.omit(WB_K562))) > 1) print(paste0("Error in RBP ", target_RBP, " cell line K562, method WB."))
  
  if(length(unique(na.omit(PCR_HepG2))) > 1) print(paste0("Error in RBP ", target_RBP, " cell line HepG2, method PCR."))
  if(length(unique(na.omit(PCR_K562))) > 1) print(paste0("Error in RBP ", target_RBP, " cell line K562, method PCR."))
}
```

## Step 5: Summarize and store to disk {-}

From the reported efficiencies for each cell line, we calculate the average of them as the reported knockdown efficiency for a particular target gene. These reported efficiencies are stored for both Western blot and qRT-PCR.

```{r}
# Write the knockdown efficiency table to disk
metadata_WB_kEff <- writeEfficiencyTable(metadata_kEff, "WB", metadata_WB_output)
metadata_PCR_kEff <- writeEfficiencyTable(metadata_kEff, "PCR", metadata_PCR_output)
```

# Results

## Knockdown Efficiency {#kEff}

In the following table, we can observe the obtained knockdown efficiencies for every studied target gene:

```{r tpm-table, echo = F}
metadata_WB_kEff <- metadata_WB_kEff %>%
  left_join(metadata_RBPs %>% select(target_gene, Category) %>% distinct(),
            by = "target_gene") %>%
  relocate(Category, .after = target_gene) %>%
  filter(!is.na(kEff_avg))

metadata_WB_kEff %>%
  select(-kEff_HepG2, -kEff_K562) %>%
  `colnames<-`(c("Target gene", "Category", "kEff [%]")) %>%
  prettier_DT(order = 3)
```

From the `r metadata_RBPs %>% pull(target_gene) %>% unique %>% length` studied projects, `r metadata_RBPs %>% left_join(metadata_WB_kEff %>% select(target_gene, kEff_avg), by = "target_gene") %>% filter(is.na(kEff_avg)) %>% pull(target_gene) %>% unique %>% length` of them had no available information for any of the two cell lines. The average knockdown efficiency is `r metadata_WB_kEff$kEff_avg %>% mean(na.rm = T) %>% round(1)`%, ranging from `r metadata_WB_kEff$kEff_avg %>% min(na.rm = T) %>% round(1)` to `r metadata_WB_kEff$kEff_avg %>% max(na.rm = T) %>% round(1)`% in some cases. In the following visualization, we represent in the X-axis the knockdown efficiency for every studied target RBP/NMD, split by their functional category:

```{r kEff-category-graph, echo = F, fig.height=9, fig.width=7.2, dpi = 300}
kEff_df <- metadata_WB_kEff %>%
  arrange(-kEff_avg) %>%
  mutate(target_gene = factor(target_gene, levels = .$target_gene),
         Category = factor(Category, levels = c("Splicing_regulation", "Spliceosome", "NMD", "Exon_junction_complex")))

ggplot(kEff_df) +
  geom_bar(aes(x = target_gene, y = kEff_avg, fill = Category), stat = "identity", color = "black", linewidth = 0.25, width = 0.8) +
  viridis::scale_fill_viridis(begin = 0.2, discrete = T, labels = category_labels, guide = guide_legend(nrow=2, byrow=TRUE)) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.02))) +
  scale_x_discrete(expand = expansion(add = c(0.7, 0.7))) +
  coord_flip(ylim = c(0, 100)) +
  labs(x = "Target shRNA knockdown gene", y = "Knockdown Efficiency [%]") +
  ggforce::facet_col(vars(Category),
                     scales = "free_y",
                     space = "free",
                     labeller = labeller(Category = category_labels)) +
  custom_gg_theme + theme(plot.title = element_text(size = 20, face = "bold"),
                         legend.title = element_text(size=10),
                         legend.text = element_text(size=8),
                         legend.key.size = unit(0.75, 'lines'))

#ggsave(file = "images/kEff_category.png", width = 183, height = 183*9/7.2, units = "mm", dpi = 300)
```

## Cell line influence

In some cases, the difference in $kEff$ between the two cell lines is not irrelevant. In the following table and graph, we show the $kEff$ for each cell line and the difference between the two:

```{r kEff-cell_line, echo = F}
metadata_WB_kEff %>%
  mutate(Diff = abs(kEff_HepG2 - kEff_K562)) %>%
  select(-kEff_avg) %>%
  `colnames<-`(c("Target gene", "Category", "HepG2 kEff [%]", "K562 kEff [%]", "Difference abs(HepG2 - K562)")) %>%
  mutate(across(where(is.numeric), round, 1)) %>%
  prettier_DT(order = 5)
```

```{r kEff-cell_line-graph, echo = F, fig.height=12, fig.width=7.2, dpi = 300}
kEff_df <- metadata_WB_kEff %>%
  mutate(Diff = abs(kEff_HepG2 - kEff_K562)) %>%
  arrange(-kEff_HepG2) %>%
  mutate(target_gene = factor(target_gene, levels = .$target_gene),
         Category = factor(Category, levels = c("Splicing_regulation", "Spliceosome", "NMD", "Exon_junction_complex")),
         type = ifelse(kEff_HepG2 > kEff_K562, "HepG2", "K562"))

ggplot(kEff_df) +
  geom_segment(aes(x = target_gene, y = kEff_HepG2, yend = kEff_K562, xend = target_gene, col = "Segment_col"), 
               linewidth = 1.5) +
  scale_color_manual(name = "Difference", labels = "", values = c("Segment_col" = "#5dc863"),
                     guide = guide_legend(override.aes = list(width = 10), keywidth = 2)) +
  ggnewscale::new_scale_color() + 
  geom_point(aes(x = target_gene, y = kEff_HepG2, color = "HepG2"), size = 1.5) + 
  geom_point(aes(x = target_gene, y = kEff_K562, color = "K562"), size = 1.5) + 
  scale_color_manual(name = "Cell line",
                    labels = c("HepG2", "K562"),
                    values = c("HepG2" = "black", "K562" = "#2181ff"),
                    guide = guide_legend(order = 1)) + 
  coord_flip(ylim = c(0, 100)) +
  ggforce::facet_col(vars(Category), 
                     scales = "free_y", space = "free", 
                     labeller = labeller(Category = category_labels)) +
  labs(x = "Target gene", y = "Knockdown efficiency [%]") + ggtitle("Knockdown Efficiency by cell line") + 
  scale_y_continuous(expand = expansion(mult = c(0, 0))) + 
  custom_gg_theme + theme(legend.key = element_blank(),
                          strip.text.x = element_text(color = "black", face = "bold", size = 11),
                          plot.title = element_text(size = 16),
                          legend.margin = margin(0, 1.5, 0, 0, unit="cm"))
#ggsave(file = "images/kEff_category_cell_line.png", width = 183, height = 183*9/7.2, units = "mm", dpi = 300)
```
